modules/hdf5_handler/h5get.cc

```diff
@@ -129,76 +122,33 @@ hid_t get_attr_info(hid_t dset, int index, bool is_dap4, DSattr_t * attr_inst_pt
-    // Note:DAP4 explicitly states that the data should be defined atomic datatype(int, string).
-    // 1-D variable length of string can also be mapped to both DAS and DDS.
-    // The variable length string class is H5T_STRING rather than H5T_VLEN,
-    // So safe here. 
-    // We also ignore the mapping of integer 64 bit for DAP2 since DAP2 doesn't
-    // support 64-bit integer. In theory, DAP2 doesn't support long double
-    // (128-bit or 92-bit floating point type), since this rarely happens
-    // in DAP application, we simply don't consider here.
-    // However, DAP4 supports 64-bit integer.
-    if ((ty_class == H5T_TIME) || (ty_class == H5T_BITFIELD)
-        || (ty_class == H5T_OPAQUE) || (ty_class == H5T_ENUM)
-        || (ty_class == H5T_REFERENCE) ||(ty_class == H5T_COMPOUND)
-        || (ty_class == H5T_VLEN) || (ty_class == H5T_ARRAY)){ 
-        
-        *ignore_attr_ptr = true;
+    *ignore_attr_ptr = check_ignored_attrs(attrid,ty_id,attr_name,is_dap4);
+    if (*ignore_attr_ptr) {
         H5Tclose(ty_id);
         return attrid;
     }
 
-    // Ignore 64-bit integer for DAP2.
-    // The nested if is better to understand the code. Don't combine
-    if (false == is_dap4) {
-        if((ty_class == H5T_INTEGER) && (H5Tget_size(ty_id)== 8)) {//64-bit int
-            *ignore_attr_ptr = true;
-            H5Tclose(ty_id);
-            return attrid;
-        }
-    }       
-
-    // Here we ignore netCDF-4 specific attributes for DAP4 to make filenetCDF-4 work
-    if (true == is_dap4 && HDF5RequestHandler::get_default_handle_dimension() == true) {
-        // Remove the nullptrTERM etc.
-        string attr_name_str(attr_name.begin(),attr_name.end()-1);
-        if(attr_name_str == "CLASS" || attr_name_str == "NAME" || attr_name_str == "_Netcdf4Dimid" 
-           || attr_name_str == "_nc3_strict" || attr_name_str=="_NCProperties" || attr_name_str=="_Netcdf4Coordinates") {
-            *ignore_attr_ptr = true;
-            H5Tclose(ty_id);
-            return attrid;
-        }
-    }
-
     hid_t aspace_id = -1;
     if ((aspace_id = H5Aget_space(attrid)) < 0) {
         string msg = "cannot get hdf5 dataspace id for the attribute ";
         string attrnamestr(attr_name.begin(),attr_name.end());
         msg += attrnamestr;
+        H5Tclose(ty_id);
         H5Aclose(attrid);
         throw InternalErr(__FILE__, __LINE__, msg);
     }
 
     // It is better to use the dynamic allocation of the array.
     // However, since the DODS_MAX_RANK is not big and it is also
-    // used in other location, we still keep the original code.
+    // used in other locations, we still keep the original code.
     // KY 2011-11-16
 
     int ndims = H5Sget_simple_extent_ndims(aspace_id);
     if (ndims < 0) {
         string msg = "cannot get hdf5 dataspace number of dimension for attribute ";
         string attrnamestr(attr_name.begin(),attr_name.end());
         msg += attrnamestr;
+        H5Tclose(ty_id);
```

Comment:
One thing you can do - not a change you need to make, but a technique that can be useful - is to make a class with a constructor and destructor that will manage the need to call multiple 'close()' functions before an exception is thrown. Those can be classes local to a file and they can help ensure that all the things are closed every time that is needed.

Reply:
Will discuss with you on this later.